doc/wol/Wake-on-LAN-HLD.md

```diff
@@ -0,0 +1,172 @@
+         +------+------+------+------+------+------+------------------------------
+      14 | 0xFF | 0xFF | 0xFF | 0xFF | 0xFF | 0xFF |
+         +------+------+------+------+------+------+        ETHERNET FRAME
+      20 | Target MAC (repeat 16 times, 96 bytes)  |           PAYLOAD
+         +-----------------------------------------+
+     116 |    Password (optional, 4 or 6 bytes)    |
+         +-----------------------------------------+
+```
+
+## CLI Design
+
+The `wol` command is used to send magic packet to target device.
+
+### Usage
+
+```
+wol <interface> <target_mac> [-b] [-p password] [-c count] [-i interval]
+```
+
+- `interface`: SONiC interface name.
+- `target_mac`: MAC address of the target device.
+- `-b`: Use broadcast MAC address instead of target device's MAC address as **Destination MAC Address in Ethernet Frame Header**.
+- `-p password`: An optional 4 or 6 byte password, in ethernet hex format or quad-dotted decimal[^3].
+- `-c count`: Count of magic packet to send. `count` must between 1 and 5. Default value is 1. This param must use with `-i`.
+- `-i interval`: Wait `interval` milliseconds between sending each magic packet. `interval` must between 0 and 2000. Default value is 0. This param must use with `-c`.
+
+### Example
+
+```
+admin@sonic:~$ wol Ethernet10 00:11:22:33:44:55
+admin@sonic:~$ wol Ethernet10 00:11:22:33:44:55 -b
+admin@sonic:~$ wol Vlan1000 00:11:22:33:44:55 -p 00:22:44:66:88:aa
+admin@sonic:~$ wol Vlan1000 00:11:22:33:44:55 -p 192.168.1.1 -c 3 -i 2000
+```
+
+## gNOI Design
+
+The gNOI service `SONiCWolService` will provide a `Wol` interface to user, which can be called to send magic packet to target device:
+
+```proto
+syntax = "proto3";
+
+package gnoi.sonic_wol;
+
+//option (types.gnoi_version) = "0.1.0";
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+service SonicWolService {
+  rpc Wol(WolRequest) returns (WolResponse) {}
+}
+
+message WolRequest {
+    string interface = 1;         // SONiC interface name
+    string target_mac = 2;        // MAC addresses in colon notation
+    optional bool broadcast = 3;  // Default false
+    optional string password = 4; // In ethernet hex format or quad-dotted decimal
+    optional int32 count = 5;     // Count of magic packet to send, must use with interval together.
+    optional int32 interval = 6;  // Wait interval milliseconds between sending each magic packet, must use with count together.
+}
+
+message WolResponse {
+    SonicOutput output = 1;
+}
+```
+
+## Test Plan
+
+### Unit Tests for `wol` CLI
+| Case Description | Expected Result |
+| :- | :- |
+| Input a valid SONiC interface name. | Parameter validation pass, send magic packet |
+| Input an invalid SONiC interface name. | Parameter validation Fail |
+| Input a valid SONiC interface name, but the interface status is not `up`. | Return `Error: interface not up` |
```

Comment:
As discussed in the community-meeting, this error can be thrown only when the input interface is ethernet-interface by validating the interface-status as UP. 

If the input interface is VLAN, then actual host can be connected on any of the VLAN-member-ports and that actual target-device connected interface status can be down at that point. But since the VLAN-interface-status can be up, user will not see any error and user might think actual target device is woken up successfully, but in reality due to transient actual interface-status down/up flap, magic-pkt might not have reached target-host. 

In this case, user initiating this cmd might not be aware they have to retry this cmd again to wake up the target-host again. May be, we can consider throwing warning, if any of the VLAN-member interface-status is down at the time of issuing this cmd and then it's upto the user to fix interface-status and reissue the cmd again.

It's better to explain, how this situation can be handled in HLD.

Reply:
Hi @rathnasabapathyv, thanks for your question. After consideration, I think if users input VLAN as the interface, then the `wol` utility will only ensure the packet is delivered to target VLAN, but not guarantee to reach each physical interface. So,
1. Users should have some monitor mechanism to check if the target device is wake up as expected. 
2. If users want to ensure the magic-packet be sent out from a physical interface, they should specify the physical interface name instead of vlan name as the `wol` CLI parameter.